### PR TITLE
Hold cs_main while calling UpdatedBlockTip() signal

### LINKED ISSUE #12978: p2p_compactblocks.py failing occasionally on master
Reported here https://github.com/bitcoin/bitcoin/pull/12629#issuecomment-381095603 by @Empact.  I've seen it now in a few different travis jobs, including:

https://travis-ci.org/bitcoin/bitcoin/jobs/366157920
https://travis-ci.org/bitcoin/bitcoin/jobs/366139851

From my initial investigation it seems like there may be an actual bug in block announcement, so I'm  opening this issue to track.

### PR DESCRIPTION
**Pull request by skeees**
Resolves #12978

### CODE REVIEW

**Comment by skeees** [comment]
@ryanofsky / other qt experts - the gui signals still get called out of cs_main - do you see any potential issues there?

**Comment by ryanofsky** [comment]
> @ryanofsky / other qt experts - the gui signals still get called out of cs_main - do you see any potential issues there?

Would have to look into it more, but if the problem is NotifyBlockTip events being sent to the GUI in the wrong order, we should determine whether one of @TheBlueMatt's PRs fixes this. Assuming one does, it's probably sufficient for this PR to just update the `NotifyBlockTip` documentation to say that notifications can arrive at the same time from multiple threads and there are no ordering guarantees (or whatever the current situation is).

**Comment by skeees** [comment]
#11856 is the one that I think you are referring to

**Comment by TheBlueMatt** [comment]
I think this PR is fine as-is, the GUI I'd be surprised if it were broken, though the mining stuff may be (but #11856 pulls it all out to using validationinterface, which should be a less-likely-to-break-things fix as it doesn't introduce new lockorders). It'd also be nice to see the no-change-callbacks go away ala a9db3dada0119c183d16627805e90c4dbca05c6a since we're touching this code anyway.

**Comment by ryanofsky** [comment]
> I think this PR is fine as-is

I think it would be good for this PR to update `NotifyBlockTip` documentation if notifications can come out of order, regardless of whether there are currently any gui bugs.

**Comment by skeees** [comment]
@ryanofsky - i'll update docs
@TheBlueMatt - will evaluate whether feasible to cherry-pick that commit into this pr

**Comment by laanwj** [comment]
> @ryanofsky / other qt experts - the gui signals still get called out of cs_main - do you see any potential issues there?

I expect the GUI doesn't care - it transfers the signals to a different thread anyway.

**Comment by jnewbery** [comment]
Looks good to me. I don't think https://github.com/bitcoin/bitcoin/commit/a9db3dada0119c183d16627805e90c4dbca05c6a needs to be pulled in - it's an orthogonal change for whether to fire notifications at all. It's useful, but no need to widen the scope of this PR.

**Comment by TheBlueMatt** [comment]
Heh, @jnewbery pointed out that a9db3dada0119c183d16627805e90c4dbca05c6a doesn't actually fix it, I was looking at the wrong callback, sorry for the noise. In any case, I think this is fine as-is, with a UI doc change and then hopefully #11856 in the future to get rid of the ui callback entirely.

**Comment by TheBlueMatt** [comment]
As for the uiInterface callback, I think its best to move it into the cs_main as well - as @laanwj points out in the GUI its immediately pushed to a queue for main thread execution so that should be fine, and the RPCNotifyBlockChange call should be cs_main-safe as it only notify_all()s and the cs_blockchange mutex is never taken with cs_main in another thread. Obviously this really sucks and risks future bugs and I'd like to see #11856 kill the uiInterface callback there entirely, but for a backport/fix, it looks pretty safe to me with current code.

File: src/validation.cpp [hunk 1]
```diff
@@ -2677,14 +2677,15 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
                 assert(trace.pblock && trace.pindex);
                 GetMainSignals().BlockConnected(trace.pblock, trace.pindex, trace.conflictedTxs);
             }
+
+            // Notify external listeners about the new tip.
+            // Enqueue while holding cs_main to ensure that UpdatedBlockTip is called in the order in which blocks are connected
+            GetMainSignals().UpdatedBlockTip(pindexNewTip, pindexFork, fInitialDownload);
```
Comment (promag) [inline src/validation.cpp]: Should add
```cpp
AssertLockHeld(cs_main);
````
to `CMainSignals::UpdatedBlockTip`?

File: src/validation.cpp [hunk 2]
```diff
@@ -2677,14 +2677,15 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
...
         // When we reach this point, we switched to a new tip (stored in pindexNewTip).
 
         // Notifications/callbacks that can run without cs_main
 
-        // Notify external listeners about the new tip.
-        GetMainSignals().UpdatedBlockTip(pindexNewTip, pindexFork, fInitialDownload);
-
         // Always notify the UI if a new block tip was connected
         if (pindexFork != pindexNewTip) {
             uiInterface.NotifyBlockTip(fInitialDownload, pindexNewTip);
```
Comment (promag) [inline src/validation.cpp]: This means that the UI can present a tip that is not the tip?

**Comment by skeees** [comment]
Consensus seems to be to update the gui callbacks in cs_main as well (until #11856 which moves these gui callbacks to the ValidationInterface) - latest commit does that and addresses @promag's other review comment

**Comment by jnewbery** [comment]
ACK b6f486c85c85e024f03941cbbeadb348ddceaaf8. Commits can be squashed.

**Comment by skeees** [comment]
Squashed

File: src/validationinterface.cpp [hunk 3]
```diff
@@ -139,6 +139,11 @@ void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason
 }
 
 void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {
+    // Dependencies exist that require UpdatedBlockTip events to be delivered in the order in which
+    // the chain actually updates. In order to ensure this, these signals should be enqueued
+    // in the same critical section where the chain is updated
+    AssertLockHeld(cs_main);
```
Comment (TheBlueMatt) [inline src/validationinterface.cpp]: Bleh, if you're gonna assert this here you should probably assert it in pretty much every signal, as they all have ordering requirements. It also just seems kinda strange to do so cause its not really a requirement of the validationinterface, its a requirement that they be ordered, not done in one specific cs (eg the mempool ones could probably be ordered by mempool.cs instead).

**Review by TheBlueMatt — COMMENTED** [review]
utACK 713c066cdc1cc4750dfa1d41149621356ff2ad04

File: src/validationinterface.cpp (hunk 3 above)
Comment (sdaftuar) [reply src/validationinterface.cpp]: I was also not loving this AssertLockHeld() for the same reason.  I'd prefer we drop it and just leave a comment explaining the ordering requirement (maybe with a comment mentioning the cs_main lock as one way to achieve it).

Comment (skeees) [reply src/validationinterface.cpp]: Agreed - updated

Comment (promag) [reply src/validationinterface.cpp]: I'm not a fan of spreading `AssertLockHeld` either but in this case the current fix is to lock and having the assert here makes that explicitly required. With just the comment a future refactor can break that requirement and reintroduce the bug.

But it was just a suggestion.

Comment (sdaftuar) [reply src/validationinterface.cpp]: Note that the `AssertLockHeld` wouldn't have prevented the caller from releasing cs_main after updating the tip, and reacquiring cs_main before the validation interface callbacks -- which would still result in violating the ordering requirements.

Comment (promag) [reply src/validationinterface.cpp]: @sdaftuar that's true and the same could be said to every `AssertLockHeld` 😛

**Comment by sdaftuar** [comment]
utACK d86edd3

**Pull request merged**